### DISCUSSION TITLE
Forcing dynamic on multiple routes/project-wide

### DISCUSSION THREAD

**Discussion by NotNite** [vercel/next.js#48989] [App Router]
Heya! Me and a friend (@slice) are working on a new project using the App Router. We've run into an issue right before deployment with caching. Source available [here](https://git.n2.pm/NotNet/gluestick/src/commit/509b4a8f424e32ca468a9fdc5aaef7db90be83e5). Explanation follows, TL;DR is bolded.

## Our problem

We have OAuth integration with Discord and GitHub in our project, so we have a route on `/oauth/discord/login/route.ts` ([link](https://git.n2.pm/NotNet/gluestick/src/commit/509b4a8f424e32ca468a9fdc5aaef7db90be83e5/src/app/oauth/discord/login/route.ts)) that redirects the user to Discord (status: 302, headers.Location = redirectUrl). The redirect URL contains an environment variable (the client ID) in it.

Next doesn't realize this is a dynamic route, though, so it caches the redirect:

```
$ cat .next/server/app/oauth/discord/login.meta | jq .
{
  "status": 302,
  "headers": {
    "location": "https://discord.com/oauth2/<rest omitted for brevity>",
    "set-cookie": "state=86992627-7074-4d43-a543-a2118e4f650e; Path=/;"    
  }
}
```

The issues we have with this specifically are:

- This service is deployed via Nix, so there are no environment variables set at build time. This means when Next goes to cache the route, the environment variables are `undefined` and the redirect URL is malformed, breaking the service.
- The `Set-Cookie` header state is hardcoded to the same value every time, meaning CSRF attacks are possible through this.

We can set the [route segment config option](https://beta.nextjs.org/docs/api-reference/segment-config#dynamic) to `force-dynamic` to fix this:

```ts
export const dynamic = "force-dynamic";

export async function GET(request: Request) {
  // ...
}
```

However, this means that we have to set this on basically every route that Next doesn't detect as dynamic.

For us, there are only two files for us that need to have this override:

```shell
$ find .next/server -name "*.meta"
.next/server/app/oauth/discord/login.meta
.next/server/app/oauth/github/login.meta
```

## The question

**Is it possible to force multiple routes to be dynamic without needing to edit every single individual file** (say, being able to regex a route or set it globally in `next.config.js`)?

This seems like a weird security hole to me, regardless of if this option exists: if we weren't using Nix for deploying this service, I wouldn't have noticed this (redirect URL wouldn't have been malformed), meaning CSRF attacks could have been possible. This isn't a smart default to set, though.

I can't think of a good way Next can protect the developer from making this mistake - maybe it could search for `process.env` usage in routes and refuse to cache them? That doesn't fix the cached UUID, though.

## Minimal example

Consider the following example, [NotNite/force-dynamic](https://github.com/NotNite/force-dynamic). I created it with `npx create-next-app@latest --experimental-app force-dynamic`, and installed `uuid` & `@types/uuid`. It has a very simple `api/test/route.ts`:

```ts
import { v4 } from "uuid";

export async function GET(request: Request) {
  return new Response(`${process.env.TEST} ${v4()}`);
}
```

We can run the server with `npm run dev`, and see that it responds:

```shell
$ TEST="hi" npm run dev

# in another terminal
$ curl -s http://localhost:3000/api/test
hi db41aa4c-16c2-44a3-91ec-21c6100861cf
$ curl -s http://localhost:3000/api/test
hi a1939dfd-7631-493c-97e0-cf01244ee890
```

Now, we can build it, and see the response is now static:

```shell
$ TEST="hi" npm run build
$ npm run start

# in another terminal
$ curl -s http://localhost:3000/api/test
hi bd203279-0392-40be-9616-290e1ede090f
$ curl -s http://localhost:3000/api/test
hi bd203279-0392-40be-9616-290e1ede090f
$ cat .next/server/app/api/test.body
hi bd203279-0392-40be-9616-290e1ede090f
```

If we build it without the environment variable (similar to our Nix setup), it becomes `undefined`, even when providing the environment variable (because it's cached):

```shell
$ npm run build
$ TEST="hi" npm run start

# in another terminal
$ curl -s http://localhost:3000/api/test
undefined 412df24f-f7d3-4c2c-8f04-49efbf035a40
$ curl -s http://localhost:3000/a
undefined 412df24f-f7d3-4c2c-8f04-49efbf035a40
$ cat .next/server/app/api/test.body
undefined 412df24f-f7d3-4c2c-8f04-49efbf035a40
```

Setting force-dynamic ([patch](https://github.com/vercel/next.js/files/11359906/force-dynamic.patch)) fixes the issue, and the `.body` file is gone like expected:

```shell
$ cat src/app/api/test/route.ts
import { v4 } from "uuid";

export const dynamic = "force-dynamic";

export async function GET(request: Request) {
  return new Response(`${process.env.TEST} ${v4()}`);
}
```

```shell
$ npm run build
$ TEST="hi" npm run start

# in another terminal
$ curl -s http://localhost:3000/api/test
hi c447cd34-c00f-4043-8047-207109ac3e69
$ curl -s http://localhost:3000/api/test
hi e9491096-55f1-4c6a-aa9a-e9bcf78b5840
$ ls .next/server/app/api/test.body
ls: cannot access '.next/server/app/api/test.body': No such file or directory
```

**Comment by slice** [comment]
@NotNite Looking at the docs, I'm assuming you can set whether Next.js considers things to be dynamic or not at the layout level. If we do this in the root layout, it should apply globally. Try this?

`src/app/layout.tsx`
```ts
 export const dynamic = 'force-dynamic';
```

This doesn't seem like a security hole to me; in fact, it appears to be intended behavior. Maybe Next.js can run heuristics on route segment names to help the developer, telling them that they probably wouldn't want to cache routes that contain `oauth`, `redirect`, etc. 

Low latency execution on the edge is one of Next.js's major priorities so it makes sense to aggressively cache things that appear static.

**Reply by pavel-krasnov** [reply]
Though the first load will be totally dynamic, when you then navigate on client and some server component is executed, its fetch requests will be cached unless you make the respective segment dynamic too.

**Comment by sravimohan** [comment]
@NotNite How did you go with this?
